packages/application/src/index.ts

```diff
@@ -327,54 +328,63 @@ export class Application<T extends Widget> {
     // Reject the promise if the plugin is not registered.
     let data = this._pluginMap.get(id);
     if (!data) {
-      return Promise.reject(new Error(`Plugin '${id}' is not registered.`));
+      throw new Error(`Plugin '${id}' is not registered.`);
     }
 
+    // Bail early if the plugin is not activated
     if (!data.activated) {
-      // Bail early if the plugin is not activated
-      return Promise.resolve([]);
+      return [];
     }
 
     if (!data.deactivate) {
-      return Promise.reject(`Plugin '${id}' does not support deactivation`);
+      throw new Error(`Plugin '${id}' does not support deactivation`);
     }
 
-    // find an optimal deactivation order for dependants
-    const dependentIds = Private.findDependants(
+    // Find the optimal deactivation order for dependents.
+    const dependentIds = Private.findDependents(
       id,
       this._pluginMap,
       this._serviceMap
     );
 
-    const dependants = dependentIds.map(d => this._pluginMap.get(d)!);
+    const dependents = dependentIds.map(d => this._pluginMap.get(d)!);
 
-    // Check all dependents
-    for (const dependant of dependants) {
-      if (!dependant.deactivate) {
-        Promise.reject(
-          `Dependent plugin ${dependant.id} does not support deactivation.`
+    // Check to make sure all dependent plugins can deactivate.
+    for (const dependent of dependents) {
+      if (!dependent.deactivate) {
+        throw new Error(
+          `Dependent plugin ${dependent.id} does not support deactivation.`
         );
       }
     }
 
-    for (const dependant of dependants) {
-      const deps = dependant.requires.concat(dependant.optional);
-      await Promise.resolve(
```

Comment:
In this version, we `await` promises sequentially, I believe

Reply:
I went for an sequential execution because the dependents are passed to the deactivate function (or maybe used within it). So it feels wrong to provide plugins that may be destroyed prior or during the deactivation call.

If we want to optimize deactivation, we should do something similar to the activation state:

Promise.all(_dependents.deactivate_).then(service.deactivate). But I did not go that path as it will implies more topology introspection to get the list of dependents for each plugin.

Reply:
That makes sense, thanks for explaining! I switched it back.